### ISSUE TITLE
NVD conditions aren't fully applied by Wazuh vulnerability detector

### ISSUE THREAD

**Issue by dariommr** [wazuh/wazuh#8426]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.1.5 | Vulnerability Detector | Manager | Packages | Container |

Hello Team,
I experienced this issue with the [CVE-2019-11043](https://nvd.nist.gov/vuln/detail/CVE-2019-11043).
The condition in the rule seems to be wrong (or at least not precise). More details:

In the event, this is shown:
```
Version: "7.2.24-1.module_el8.2.0+313+b04d0a66"
Condition: "Package less than 7.3.5-5.module+el8.1.0+4560+e0eee7d6"
Rationale: "In PHP versions 7.1.x below 7.1.33, 7.2.x below 7.2.24 and 7.3.x below 7.3.11 in certain configurations of FPM setup it is possible to cause FPM module to write past allocated buffers into the space reserved for FCGI protocol data, thus opening the possibility of remote code execution."
```
But if we check the NVD bulletin, we can see clearly that the conditions are several:
![image](https://user-images.githubusercontent.com/37050249/116453571-68ce0f80-a835-11eb-8858-a65a9058e2a9.png)

The version reported, seems to be excluded from the Vulnerability Scope.
Can you please check this?
Thank you in advance.

**Comment by dariommr** [comment]
Here is the alert in JSON Format:
```
{
  "_index": "wazuh-alerts-4.x-v1-2021.04.29-000022",
  "_type": "_doc",
  "_id": "q-DhP3kBUyCHdQc-n4-L",
  "_version": 1,
  "_score": null,
  "_source": {
    "input": {
      "type": "log"
    },
    "cluster": {
      "node": "wazuh-manager-worker-1",
      "name": "wazuh"
    },
    "agent": {
      "ip": "1.1.1.1",
      "name": "agent_name",
      "id": "1"
    },
    "manager": {
      "name": "wazuh-manager-worker-1"
    },
    "data": {
      "vulnerability": {
        "severity": "Critical",
        "package": {
          "condition": "Package less than 7.3.5-5.module+el8.1.0+4560+e0eee7d6",
          "name": "php-xml",
          "version": "7.2.24-1.module_el8.2.0+313+b04d0a66",
          "architecture": "x86_64"
        },
        "references": [
          "https://github.com/neex/phuip-fpizdam",
          "https://bugs.php.net/bug.php?id=78599",
          "https://usn.ubuntu.com/4166-1/",
          "https://usn.ubuntu.com/4166-2/",
          "https://www.debian.org/security/2019/dsa-4553",
          "https://www.debian.org/security/2019/dsa-4552",
          "https://support.f5.com/csp/article/K75408500?utm_source=f5support&amp;utm_medium=RSS",
          "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/T62LF4ZWVV7OMMIZFO6IFO5QLZKK7YRD/",
          "https://security.netapp.com/advisory/ntap-20191031-0003/",
          "https://access.redhat.com/errata/RHSA-2019:3287",
          "https://access.redhat.com/errata/RHSA-2019:3286",
          "https://access.redhat.com/errata/RHSA-2019:3299",
          "https://access.redhat.com/errata/RHSA-2019:3300",
          "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/3W23TP6X4H7LB645FYZLUPNIRD5W3EPU/",
          "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/FSNBUSPKMLUHHOADROKNG5GDWDCRHT5M/",
          "http://lists.opensuse.org/opensuse-security-announce/2019-11/msg00011.html",
          "https://access.redhat.com/errata/RHSA-2019:3724",
          "https://access.redhat.com/errata/RHSA-2019:3735",
          "https://access.redhat.com/errata/RHSA-2019:3736",
          "https://www.synology.com/security/advisory/Synology_SA_19_36",
          "http://lists.opensuse.org/opensuse-security-announce/2019-11/msg00014.html",
          "https://support.apple.com/kb/HT210919",
          "https://seclists.org/bugtraq/2020/Jan/44",
          "http://seclists.org/fulldisclosure/2020/Jan/40",
          "https://access.redhat.com/errata/RHSA-2020:0322",
          "http://packetstormsecurity.com/files/156642/PHP-FPM-7.x-Remote-Code-Execution.html",
          "https://nvd.nist.gov/vuln/detail/CVE-2019-11043",
          "https://access.redhat.com/security/cve/CVE-2019-11043"
        ],
        "cve_version": "4.0",
        "assigner": "security@php.net",
        "cwe_reference": "CWE-787",
        "title": "CVE-2019-11043 affects php-xml",
        "rationale": "In PHP versions 7.1.x below 7.1.33, 7.2.x below 7.2.24 and 7.3.x below 7.3.11 in certain configurations of FPM setup it is possible to cause FPM module to write past allocated buffers into the space reserved for FCGI protocol data, thus opening the possibility of remote code execution.",
        "advisories_ids": [
          "RHSA-2019:3286",
          "RHSA-2019:3287",
          "RHSA-2019:3299",
          "RHSA-2019:3300",
          "RHSA-2019:3724",
          "RHSA-2019:3735",
          "RHSA-2019:3736",
          "RHSA-2020:0322",
          "RHSA-2020:2835"
        ],
        "cve": "CVE-2019-11043",
        "bugzilla_references": [
          "https://bugzilla.redhat.com/show_bug.cgi?id=1766378"
        ],
        "updated": "2020-08-18",
        "cvss": {
          "cvss2": {
            "base_score": "7.500000",
            "vector": {
              "integrity_impact": "partial",
              "confidentiality_impact": "partial",
              "availability": "partial",
              "attack_vector": "network",
              "access_complexity": "low",
              "authentication": "none"
            }
          },
          "cvss3": {
            "base_score": "9.800000",
            "vector": {
              "user_interaction": "none",
              "integrity_impact": "high",
              "scope": "unchanged",
              "confidentiality_impact": "high",
              "availability": "high",
              "attack_vector": "network",
              "access_complexity": "low",
              "privileges_required": "none"
            }
          }
        }
      }
    },
    "rule": {
      "firedtimes": 16,
      "mail": true,
      "tsc": [
        "CC7.1",
        "CC7.2"
      ],
      "pci_dss": [
        "11.2.1",
        "11.2.3"
      ],
      "level": 13,
      "description": "CVE-2019-11043 affects php-xml",
      "groups": [
        "vulnerability-detector"
      ],
      "id": "23506",
      "gdpr": [
        "IV_35.7.d"
      ]
    },
    "@index_name": "wazuh-alerts",
    "location": "vulnerability-detector",
    "id": "1620274419.49334127",
    "decoder": {
      "name": "json"
    },
    "timestamp": "2021-05-06T04:13:39.663Z"
  },
  "fields": {
    "timestamp": [
      "2021-05-06T04:13:39.663Z"
    ]
  },
  "highlight": {
    "rule.id": [
      "@kibana-highlighted-field@23506@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1620274419663
  ]
}
```

**Comment by chemamartinez** [comment]
Hi @dariommr and team,

I have reviewed this case more in depth. Related to CVE-2019-11043 we have this information in the vulnerability database for RHEL8:

```
select * from VULNERABILITIES where CVEID="CVE-2019-11043" and PACKAGE="php-xml";
CVE-2019-11043|RHEL8||php-xml|less than|7.3.5-5.module+el8.1.0+4560+e0eee7d6|0|0|1041
CVE-2019-11043|RHEL8||php-xml|less than|7.2.11-4.module+el8.1.0+4555+f5cb8e18|0|0|1047
```

For this CVE and package, they released a fix in two different minors. When this appears in the OVAL feed, Vulnerability Detector is not ready to take this into account, it can report that a package higher than 7.2.11 but less than 7.3.5 is vulnerable by matching the first entry of the DB:
```
2021/05/09 16:57:07 wazuh-modulesd:vulnerability-detector[6324] wm_vuln_detector.c:2042 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5458): Package 'php-xml' inserted into the vulnerability 'CVE-2019-11043'. Version (7.2.24-1.module_el8.2.0+313+b04d0a66) 'less than' '7.3.5-5.module+el8.1.0+4560+e0eee7d6' (feed 'OVAL').
```

After that, the NVD scan doesn't report this vulnerability. However, the OVAL criteria win for this case due to the vendor of the package is `CentOS`, so the alert is reported finally.
```
2021/05/09 16:57:26 wazuh-modulesd:vulnerability-detector[6324] wm_vuln_detector_nvd.c:2532 at wm_vuldet_check_generic_package(): DEBUG: (5460): Package 'php' not vulnerable to 'CVE-2019-11043'. Version (7.2.24-1.module_el8.2.0+313+b04d0a66) not 'less than' '7.2.24' (feed 'NVD').
2021/05/09 16:58:37 wazuh-modulesd:vulnerability-detector[6324] wm_vuln_detector.c:1526 at wm_vuldet_send_cve_report(): DEBUG: (5468): The 'php-xml' package (7.2.24-1.module_el8.2.0+313+b04d0a66) from agent '001' is vulnerable to 'CVE-2019-11043'. Condition: 'Package less than 7.3.5-5.module+el8.1.0+4560+e0eee7d6'
```

**First approach**

I have reviewed the Vulnerability Detector code in charge of detecting false positives, the function `wm_vuldet_linux_rm_oval_not_affected_packages` evaluates the OVAL results against the NVD to double-check for false positives, but only for `Unfixed` packages.

https://github.com/wazuh/wazuh/blob/f0ec8033d164692ccc5b04a64a5eeabc93890fa8/src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c#L1724

One possible solution would be to evaluate all the vulnerable packages for the OVAL against the NVD (not only the unfixed but the ones that have a fixed version). With this solution the vulnerability is no longer reported:

```
2021/05/09 17:05:46 wazuh-modulesd:vulnerability-detector[7281] wm_vuln_detector.c:2042 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5458): Package 'php-xml' inserted into the vulnerability 'CVE-2019-11043'. Version (7.2.24-1.module_el8.2.0+313+b04d0a66) 'less than' '7.3.5-5.module+el8.1.0+4560+e0eee7d6' (feed 'OVAL').
2021/05/09 17:06:25 wazuh-modulesd:vulnerability-detector[7281] wm_vuln_detector_nvd.c:2532 at wm_vuldet_check_generic_package(): DEBUG: (5460): Package 'php' not vulnerable to 'CVE-2019-11043'. Version (7.2.24-1.module_el8.2.0+313+b04d0a66) not 'less than' '7.2.24' (feed 'NVD').
2021/05/09 17:07:43 wazuh-modulesd:vulnerability-detector[7281] wm_vuln_detector.c:1755 at wm_vuldet_linux_rm_oval_not_affected_packages(): DEBUG: (5462): Package 'php-xml' not vulnerable to 'CVE-2019-11043' since it is not affected (feed 'NVD').
```

However, this change affects other packages that are not false positives but the vulnerability doesn't match the NVD for different reasons (e.g. standard naming issues)

**Proposed solution**

Another solution pending to be tested is to modify the OVAL scan to realize when a vulnerability and package have more than one entry in the vulnerabilities database, in which case it should match all of them to be vulnerable.

**Comment by alromeros** [comment]
**Devel branch**: https://github.com/wazuh/wazuh/tree/8426-fix-false-positives-in-OVAL-scan

**Comment by alromeros** [comment]
**Pull Request:** https://github.com/wazuh/wazuh/pull/8883

**Comment by chemamartinez** [comment]
Fixed at #8883. Thanks, @alromeros and @dariommr!
